cgd_tx_eff/src/edu/ohsu/compbio/txeff/tx_eff_annotate.py

```diff
@@ -0,0 +1,55 @@
+'''
+Created on Aug. 12, 2024
+
+@author: pleyte
+'''
+
+import logging
+import os
+
+from biocommons.seqrepo import SeqRepo
+
+from edu.ohsu.compbio.txeff.util import chromosome_map
+from edu.ohsu.compbio.txeff.variant_transcript import VariantTranscript
+
+
+class TxEffAnnotate(object):
+    '''
+    Add variant annotations to each variant transcript.  
+    '''
+    def __init__(self):
+        '''
+        Constructor
+        '''
+        # Set up a logger for this class
+        self.logger = logging.getLogger(__name__)
+        
+        # Setup the SeqRepo utility that will be used for the Reference Context annotation
+        if os.environ.get('HGVS_SEQREPO_DIR') is None:
+            raise ValueError("The HGVS_SEQREPO_DIR environment variable is not defined")
+        else:
+            self.logger.info(f"Using SeqRepo {os.environ.get('HGVS_SEQREPO_DIR')}")
+            self.sr = SeqRepo(os.environ.get('HGVS_SEQREPO_DIR'))
+        
+        self.REFERENCE_CONTEXT_LENGTH_PER_SIDE = 10
+        
+    def annotate(self, transcripts: list):
+        '''
+        Add annotations to each variant transcript
+        '''
+        self.logger.info("Applying Reference context annotations")
+        for transcript in transcripts:
+            transcript.reference_context = self._get_reference_context(transcript)
+
+    def _get_reference_context(self, transcript: VariantTranscript):
+        '''
+        Look up the reference sequence that surrounds a variant. 
+        '''
+        refseq_chromosome = chromosome_map.get_refseq(transcript.chromosome)
+        context_start = transcript.position - self.REFERENCE_CONTEXT_LENGTH_PER_SIDE
+        if context_start < 0:
+            context_start = 0
+            
+        context_stop = transcript.position + self.REFERENCE_CONTEXT_LENGTH_PER_SIDE - 1
```

Comment:
I believe to match the +/- 10 behavior we saw in Oncotator, we need 
`context_stop = transcript.position + len(transcript.ref) + self.REFERENCE_CONTEXT_LENGTH_PER_SIDE`

The slicing operator (:) is already begin-inclusive, end-exclusive, so no -1 is necessary. I think the behavior we see in Oncotator is 10 before the start, through 10 after the end of the ref. Meaning for a single nucleotide, we would get a 21 base sequence like so: `AAAAAAAAAAGCCCCCCCCCC` Here are some test cases you can play around with in the interpreter:
```
"TAAAAAAAAAAGCCCCCCCCCCT"[11 - 10 : 11 + 1 + 10]
"TAAAAAAAAAAGGGCCCCCCCCCCT"[11 - 10 : 11 + 3 + 10]
```
For both of these I am expecting 10 As, the central Gs, then 10 Cs. No Ts anywhere, the first case length 21, the second is length 23.

Reply:
yes, i forgot to add in the length of the original reference.  

My goal was to come up with a start and end that would result in the greatest number of matches with the oncotator values we have in CGD right now. Using these values for start and stop I get an 86% match out of 105,163 comparisons:
* context_start = position - length_per_side - 1
* context_stop = position + len(reference) + length_per_side - 1

There's no reason we need to match the oncotator values though. So we could simplify this and drop the -1's

Reply:
In that case, the context_start above is also missing a -1, but where do these -1s come from? Does the position need to be adjusted by 1 because of the method of lookup or something?

Reply:
I can approve what you have shown here in the latest commit.

* context_start = position - length_per_side - 1
* context_stop = position + len(reference) + length_per_side - 1

The -1s are due to different start positions, and the rest of the math is what I'd expect. So for a SNP at position 100, we'd get

start = 100 - 10 - 1 = 89
stop = 100 + 1 - 10 - 1 = 110

Inclusive 89-110 gives us 21 bases, which is what we most commonly see from Oncotator.